test/unit/BPool/BPool_JoinPool.t.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {BPoolBase} from './BPoolBase.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import {BNum} from 'contracts/BNum.sol';
+import {IBPool} from 'interfaces/IBPool.sol';
+
+contract BPoolJoinPool is BPoolBase {
+  uint256 public poolAmountOut = 10e18;
+  // Assets Under Management
+  uint256 public tokenAUM = 10e18;
+  uint256 public secondTokenAUM = 30e18;
+
+  // when minting n pool shares, enough amount X of every token t should be provided to statisfy
+  // Xt = n/BPT.totalSupply() * t.balanceOf(BPT)
+  // in this scenario n = 10, totalSupply = 100
+
+  // for t = token -> t.balanceOf(BPT) = tokenAUM = 10
+  // therefore Xt = 10/100*10 = 10
+  uint256 public requiredTokenIn = 1e18;
+  // for t = secondToken -> t.balanceOf(BPT) = secondTokenAUM = 30
+  // therefore Xt = 10/100*30 = 30
+  uint256 public requiredSecondTokenIn = 3e18;
+  // put another way, if the current 100 shares represent 30 secondToken, then
+  // to mint another 10, user should provide 3 secondToken so ratio stays the same.
+
+  function setUp() public virtual override {
+    super.setUp();
+    bPool.set__finalized(true);
+    // simulate expected finalize outcome
+    bPool.call__mintPoolShare(INIT_POOL_SUPPLY);
+    bPool.call__pushPoolShare(deployer, INIT_POOL_SUPPLY);
```

Comment:
i totally prefer a `deal(address(bPool), poolShareReceiver, INIT_POOL_SUPPLY, true);` for this (so it depends less on `mintPoolShare` and `pushPoolShare`.

actually, we don't care who the receiver is, so if we just `deal` it to the pool we won't need to transfer it later to do the `call_burnPoolShare` (so `poolShareReceiver = pool` makes sense)